### PR TITLE
Let LoadControl describe one file's worth of a spanned snapshot

### PR DESCRIPTION
**Pull request by apontzen**
## Motivation

`GadgetHDFSnap` and its derivatives (`SwiftSnap`, Gadget-4, Arepo, pkdgrav3, subfind) read spanned HDF5 files by walking `LoadControl.iterate_with_interrupts` with a callback that fires when the disk cursor crosses a file boundary. Two consequences:

- Reading file *N* depends on having walked files *0..N-1*, because the read cursor and the memory write position are both accumulated along the way.
- The boundary handling inside `iterate_with_interrupts` has to split a chunk's index arrays in two (via `concatenate_indexing` plus an explicit `copy.copy`). That splitting is where #955 came from.

This is a pure refactor: no behaviour change, no new dependencies, no threads. It removes the stateful cursor, and makes each file independently describable — which is the prerequisite for reading spanned files concurrently later on.

## What changed

**`LoadControl.iterate_within(family, disk_lo, disk_hi)`** — new generator yielding the same `(readlen, disk_index, mem_index)` triples as `iterate`, but restricted to one contiguous window of a family's on-disk extent, i.e. one file of a spanned snapshot.

The window is independent of its predecessors because of one observation: a family's selected ids are ascending, so the number of them below `disk_lo` is *simultaneously* where this window's selection begins within the id list and how many memory slots all preceding windows consume between them. A single `np.searchsorted` therefore supplies both, in place of a walk. Chunks are then generated inside the window and never straddle a boundary, so no index splitting is needed at all. `mem_index` is relative to the start of the family's in-memory data.

**`HDFArrayLoader.load_arrays`** now iterates per file: each file is resolved, iterated and filled on its own terms. `_HDFFileIterator`, the interrupt callback and the shared `particle_offset` cursor are all gone, and the loop is 20 lines shorter. The dataset lookup is deliberately lazy — with partial loading most files can contribute nothing, and each lookup is a metadata round trip that is expensive on a parallel filesystem.

**`iterate` and `iterate_with_interrupts` are untouched.** `grafic._read_grafic_file` still uses the latter, for skipping Fortran record header/footer pairs *within* a single file — a genuinely different use that this change does not affect.

## Tests

`LoadControl` has had no direct unit tests until now, only indirect coverage through the readers. `tests/chunk_test.py` adds 43.

The central test uses the old code path as an oracle: it builds the `{disk position: memory index}` mapping twice — once by walking `iterate_with_interrupts` with a file-boundary callback exactly as `gadgethdf` used to, and once by calling `iterate_within` per file — and requires them to agree. It runs over eight take patterns (full load, contiguous block starting mid-file, random 1%, random 60%, strided, single particle, first-and-last, empty) against a deliberately awkward layout `[2200, 0, 300, 1800, 1000, 700]` that contains an empty file, a file far shorter than a chunk, and one exactly a chunk long. Each mapping is also checked independently against the identity between selected ids and memory positions, so the tests do not merely confirm that the two paths agree with each other.

The rest cover ground that was previously untested: chunk sizes never exceeding `max_chunk`, `iterate_within` reconstructing what `iterate` produces, `mem_index` restarting at zero per family (which `gadgethdf` relies on, laying several load-control families end-to-end into one pynbody family array), multi-family memory slices, `slice` takes being normalised against the disk length, and the `ValueError`s for negative-step slices and non-ascending or duplicated ids.

Full suites run green — `chunk`, `gadgethdf`, `swift`, `subfindhdf`, `pkdgravhdf`, `subfindhdf_gadget4`, `split_swift_snapshot`, `grafic`, `tipsy`, `nchilada`: 192 passed.

## Note for reviewers

Chunk boundaries move: they previously sat on a global per-family `max_chunk` grid and were cut by file boundaries, and now sit on a per-file grid. The disk-to-memory mapping is unchanged (that is what the oracle test pins down), but read *granularity* differs — a file shorter than `max_chunk` becomes one short read rather than part of a larger one. Nothing asserts on read sizes; the one existing use of `_max_buf` in the tests is a data assertion about a `take` straddling a chunk boundary, and it passes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01TY3MQjiggYaMuMKinBoULf

---
_Generated by [Claude Code](https://claude.ai/code/session_01TY3MQjiggYaMuMKinBoULf)_

### CODE REVIEW

File: pynbody/chunk/__init__.py [hunk 1]
```diff
@@ -389,3 +391,88 @@ def iterate(self, families_on_disk: list[family.Family], families_in_memory: lis
...
+        memory_index : slice | None
+            Slice to write into memory, relative to the start of the family, or None if ``buffer_index`` is None
+
+        """
+
+        if family not in self._disk_family_slice:
+            raise ValueError(f"Family {family} is not present on disk")
+
+        if disk_hi < disk_lo:
+            raise ValueError("disk_hi must not be less than disk_lo")
```
[the remark this answers is not in the document]
Comment (apontzen) [reply pynbody/chunk/__init__.py]: Good catch — both failure modes reproduce. Fixed in 5c184a9.

Confirming them first, against the previous head:

- `iterate_within(dm, -5, 10)` on a full load yielded `mem_index = slice(-5, 10)`, which indexes backwards from the end of the destination array rather than erroring.
- `iterate_within(dm, 0, 5000)` on a 1000-particle family instructed the caller to read 5000 entries.

Both endpoints are now checked against the family-relative length:

```python
disk_slice = self._disk_family_slice[family]
family_length = disk_slice.stop - disk_slice.start
if not 0 <= disk_lo <= disk_hi <= family_length:
    raise ValueError(...)
```

which subsumes the ordering check that was there. Tests cover reversed, negative-start, past-the-end and wholly-past-the-end windows, for both a full load and a partial one, plus the boundary case of the whole family as a single window. The ten affected suites still pass (202 tests), so nothing the readers legitimately pass is now rejected — `gadgethdf` derives its windows from `_file_interrupt_points`, whose last entry equals the family's on-disk length exactly.

---
_Generated by [Claude Code](https://claude.ai/code)_
